Assets/Scripts/Map/Map.cs

```diff
@@ -0,0 +1,57 @@
+using UnityEngine;
+
+namespace Assets.Scripts.Map
+{
+    public class Map
+    {   
+        internal static IMapState optionsState = new OptionsMapState();
+        internal static IMapState activeState = new ActiveMapState();
+        internal static IMapState menuState = new MenuMapState();
+        internal static IMapState gameEndState = new GameEndMapState();
+        internal static IMapState levelUpState = new LevelUpMapState();
+        internal IMapState state = activeState;
+        internal IMapState previousState = activeState;
+        private IMapController _mapController;
+        public Map(IMapController mapController)
+        {
+            _mapController = mapController;
+        }
+
+        public void Menu() { state.Menu(this); }
```

Comment:
```suggestion
        public void Menu() { state.Menu(_mapController); }
```
Хотя может быть внутрь `state.Menu` надо передавать не только `IMapController`, а еще и другой интерфейс, который надо  описать, а в этом классе реализовать. Реализовать можно особым "скрытным" образом, не помню точно как это называется. Пример:

``` 
public interface IMapChange // это я назвал для примера, первое что пришло в голову
{
  void SetActive();
  void SetMenu();
  void SetLevelUp();
  void SetOptions();
  void SetGameEnd();
}

public class Map: IMapState, IMapChange
{
...
    IMapChange.SetActive() {...} // нет слова public
...
}
```
Метод `IMapChange.SetActive` не виден снаружи, но `this` легко преобразуется в `IMapChange`

Этим всем мы боремся с нарушением Interface segregation principle в этом классе "https://en.wikipedia.org/wiki/Interface_segregation_principle", в конце должны будут уйти все эти `internal` методы

Reply:
Сделал так, что теперь во все методы IMapState передаётся не только Map, но и IMapController с логикой. Так что когда методу состояния нужно что-то сделать вне Map, оно обращается к логике, а когда нужно только поменять стадию или запомнить предыдущую - вызываются методы от Map.

Reply:
Вместо выделения логики смены состояния в отдельный интерфейс и класс, просто переделал эту логику и сильно уменьшил количество кода.